### ISSUE TITLE
Champion "Static Delegates"

### ISSUE THREAD

**Issue by tannergooding** [dotnet/csharplang#302]
Superseded by https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md

- [x] Proposal added: [proposals/static-delegates.md](https://github.com/dotnet/csharplang/blob/master/proposals/static-delegates.md)
- [x] Discussed in LDM (https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-05-16.md#static-delegates)
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

**Comment by tannergooding** [comment]
FYI. @gafter. I think this one was skipped over after the proposal was merged. It is being championed by @jaredpar

**Comment by AlgorithmsAreCool** [comment]
@jaredpar I took a look at the proposal and maybe this is beyond my understanding, but if this is geared at  interop i'm not really clear what the difference or synergy between this and an `extern` function is. Maybe that isn't even a sensible question...

**Comment by DavidArno** [comment]
Could you please add some examples of how this might be used, and it's limitations? For example, in the proposal, it's stated that "Static Delegates would not work with existing APIs that use regular delegates". I'm reading that as I couldn't use this to improve the performance of:
```cs
int Foo(int p, Func<int, int> f) => f(p);
```
But could static delegates be used to provide a better performing version of `Foo`? If so, what would the signature look like?

**Comment by tannergooding** [comment]
@AlgorithmsAreCool, think of this as a type of `value delegate` (in the same vein as `ValueTuple` compared to `Tuple`).

It can be used for interop, but it can also be used as a type of lightweight delegate.

**Comment by tannergooding** [comment]
@DavidArno. It wouldn't be useable as a drop in replacement in the same way that you can't just use `ValueTuple` anywhere that you can use `Tuple`. One always exists on the heap, and the other could potentially exist on the stack.

You could convert from one to the other, so it could be 'used', but it would incur the cost of the allocation/etc.

**Comment by AlgorithmsAreCool** [comment]
Could you provide a little code snippet showing the creation, passing (to a function) and invocation of a static delegate?

Also why is it "static". Wouldn't `struct delegate` be a better moniker?

**Comment by tannergooding** [comment]
`static delegate` was just the name @jaredpar gave them in the original issue: https://github.com/dotnet/csharplang/issues/80. But, as everyone knows, computer programming only has two hard problems ;)

**NOTE:** The LDM could decide to do this entirely differently, this is just how I imagine it 😄 

As for declaration, I imagine exactly like a regular delegate, but prefixed with `static`:
```C#
static delegate T ValueFunc<T>();

static delegate IntPtr WNDPROC(
    [In] IntPtr hWnd,
    [In] uint uMsg,
    [In] IntPtr wParam,
    [In] IntPtr lParam
);
```

For creation, assignment, and passing, it would work just like a regular delegate:
* `ValueFunc<int> valueFunc = Test;`
* `var result = valueFunc();`
```C#
class TestClass
{
    ValueFunc<int> _valueFunc;

    public TestClass(ValueFunc<int> valueFunc)
    {
        valueFunc();
    }
}
```

You could not do multiple assignment, so `valueFunc += Method;` and `valueFunc -= Method;` would result in a compilation error.

**Comment by AlgorithmsAreCool** [comment]
Ah, all is made clear. The static was throwing me off. 

I like this and would use it.

**Comment by Joe4evr** [comment]
So reading Jared's main post on the subject in the other thread:
> 1. It can only bind to `static` methods.
> 2. Under the hood it is represented as a `struct` that has a single member of type `IntPtr`.

I guess this also precludes use as closures?

**Comment by tannergooding** [comment]
@Joe4evr. Possible, however just because it only binds to static methods, doesn't mean it can't access instance fields, it just means you would need to be explicit about passing in `this`.

Such as:
```C#
static delegate void ValueAction<T>(T obj);

public class MyClass
{
    private int _someField;

    public static void SomeMethod(MyClass @this)
    {
        Console.WriteLine(@this._someField);
    }

    public void Method1()
    {
        ValueAction<MyClass> x = SomeMethod;
        x(this);
    }
}
```

So lets take some code you might have today:
```C#
public class MyClass
{
    public void Method1()
    {
        int someLocal = 1;
        ValueAction x = static () => Console.WriteLine(someLocal);
        x();
    }
}
```

This gets translated (roughly) into:
```C#
public class MyClass
{
    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_0
    {
        public int someLocal;

        internal void <Method1>b__0()
        {
            Console.WriteLine(someLocal);
        }
    }

    public void Method1()
    {
        var CS$<>8__locals0 = new <>c__DisplayClass0_0();
        CS$<>8__locals0.someLocal = 1;

        Action x = CS$<>8__locals0.<Method1>b__0;
        x();
    }
}
```

You then might imagine the same being supported with static delegates:
```C#
public class MyClass
{
    public void Method1()
    {
        int someLocal = 1;
        ValueAction x = static () => Console.WriteLine(someLocal);
        x();
    }
}
```

The hard part here, is that the user declares `ValueAction`, but in order to pass in the captured state we need `ValueAction<T>`. So, provided an answer is given for the required transformation here, it should be possible.

At least without delving too much into the finer details, I could imagine something like the following working:
```C#
static delegate void ValueAction();

public class MyClass
{
    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_0
    {
        public static ThreadLocal<<>c__DisplayClass0_0> threadData;
        public int someLocal;

        internal static void <Method1>b__0()
        {
            var data = threadData.Value;
            Console.WriteLine(data.someLocal);
        }
    }

    public void Method1()
    {
        var CS$<>8__locals0 = new <>c__DisplayClass0_0();
        CS$<>8__locals0.someLocal = 1;

        ValueAction x = <>c__DisplayClass0_0.<Method1>b__0;
        <>c__DisplayClass0_0.threadData.Value = CS$<>8__locals0.someLocal;
        x();
    }
}
```

I think the primary issue with the above approach would be memory overhead for stray threads, but I honestly haven't put too much thought into it right now 😄

**Comment by tannergooding** [comment]
Looks like this was discussed briefly in LDM here: https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-05-16.md#static-delegates

**Comment by Neme12** [comment]
So if an API takes a "static delegate" then I can't capture any extra state or parameters? This is the problem with function pointers in C - you can't pass anything extra and it forces you to use global variables. You can't pass any extra `this` if it's not in the signature.

**Comment by HaloFour** [comment]
@Neme12 

> So if an API takes a "static delegate" then I can't capture any extra state or parameters?

Not via closures.  The point is to be as light-weight as function pointers, which precludes any of the signature rewriting or allocations that requires.  But if you wanted to provide a mechanism to pass state you could design your API to accept that state explicitly and then return that within the function pointer as an argument, similar to how events/callbacks were often designed prior to closures.

**Comment by Neme12** [comment]
But the problem is the API won't be designed that way because they will not know what I need. This is why globals are used everywhere in C and it's terrible. This problem was even solved with C++ and they don't have to have allocations or let alone a garbage collector and it's efficient. Can't we do better?

**Comment by HaloFour** [comment]
@Neme12 

Someone with better knowledge of C++ lambdas could correct me, but IIRC you can't actually return the a lambda created in a method without some kind of allocation (as they are on the stack) and in the case of by-reference captures the onus is on the developer to ensure that the lambda is not somehow invoked beyond the life of any of the captured values.  Any capturing lambda has to be converted into an unnamed functor-like type.

**Comment by jaredpar** [comment]
> This problem was even solved in C++ without the need for allocations, let alone a garbage collector.

C++ solves this with lambdas by allocating the state on the stack and passing it by value (or reference depending on the lambda declaration). C# can take the exact same approach. State passing by value can be accomplished via structs. This is actually how local functions work: state capture without the need for allocations. 

Today though in C# there are two sources of heap allocations: the closure and the delegate. Once delegates are static the heap allocation can be removed hence there will be a reason to focus on the closure allocation.

**Comment by 4creators** [comment]
> Once delegates are static

@jaredpar Any thoughs on timing of implementation? X.0 candidate looks like no candidate at all.

**Comment by jaredpar** [comment]
@4creators 

> Any thoughs on timing of implementation? X.0 candidate looks like no candidate at all.

That just means essentially "it would only be done in a major release and we'll figure out which one later".

The complication with static delegates is that it requires runtime changes. Imagine this scenario:

- AppDomain 1: Loads Util.dll and creates a static delegate to a method inside it. 
- AppDomain 1: then passes that static delegate to AppDomain 2. A static delegate is just implemented essentially as an IntPtr and that's serializable. 
- AppDomain 2 tries to unload AppDomain 1

If that last operation is allowed to succeed then now you have a type safe delegate that points to unloaded memory. Oops 😦 

This needs to be addressed at some level in the runtime: either by preventing the AppDomain from being unloaded or by disallowing the delegate to be passed between AppDomain instances. In general our default is to pin language changes that depend on runtime changes to major releases. 

I think it has a decent shot at 8.0 though.

**Comment by 4creators** [comment]
> I think it has a decent shot at 8.0 though.

This are good news 😄

**Comment by TrunksHub** [comment]
First off, what a great proposal. Can't wait for this to be implemented.  
One particular use case I'm looking forward to is having arrays of void pointers that can be casted to their appropriate static delegate type, with the delegates themselves stored in unmanaged memory.  
Would be great if someone could confirm that this would be possible, but I guess so, seeing that static delegates would be implemented as pure value structs.

**Comment by tannergooding** [comment]
@TrunksHub, a static delegate (at least based on the current proposal) would be blittable, so it would incur no marshalling overhead when using them with unmanaged code.

**Comment by tannergooding** [comment]
I wasn't particularly busy today so I started working on a prototype: https://github.com/tannergooding/roslyn/tree/features/value-delegates

I got far enough (with very rough code) to be able to recognize `static delegate` and emit a basic synthesized struct in the form of:
```C#
internal struct A
{
  public readonly IntPtr _method; // Might be able to just emit this as a `method ptr` field (rather than a `native int`), as it includes the signature and can be loaded/recognized directly

  public void Invoke() // This gets the signature of the declared static delegate and exists for other languages to be able to work with the static delegate type, it also makes it easier to prototype
  {
    // ISSUE: unable to decompile the method.
  }

  public A(IntPtr method)
  {
    this._method = method;
  }
}
```

The compiler doesn't look like it has any support for emitting `StandAloneSig` entries for methods today (or I am just missing it completely) and that is something that will need to be unblocked before any use of `calli` can happen.

I might take a look at this again, in a couple of weeks, when I get back from vacation.

**Comment by tannergooding** [comment]
I got the prototype (https://github.com/tannergooding/roslyn/tree/features/value-delegates) working such that `Invoke` works now. Assignment still isn't possible, but it works for interop code right now (at least with the trivial tests, and assuming default calling convention).

The compiler changes "work", but are properly not "correct" (in that they would not pass code-review).

Will work on getting assignment "working" next.

**Comment by jnm2** [comment]
I was just thinking about value delegates today while writing a method that constrains `TDelegate : class`. I wonder if value delegates will end up changing the way people think about constraining that.
Maybe we need `where TDelegate : SDelegate`... 😈

**Comment by jkotas** [comment]
> it works for interop code right now (at least with the trivial tests, and assuming default calling convention)

Note that the static delegate needs to have *unmanaged* calling convention in order for it to work for interop - I have not found it in your change.

Managed function pointers are not interchangeable with unmanaged function pointers. If you call managed function via unmanaged function pointer, or vice versa, it will lead to hangs, crashes and data corruptions.

**Comment by tannergooding** [comment]
> Note that the static delegate needs to have unmanaged calling convention in order for it to work for interop 

Right now it is basically "falling out" (for interop code). The `SynthesizedStaticDelegateInvokeMethod` type says it has a calling convention of `ThisCall`, so the `calli` gets emitted as `calli unmanaged thiscall`.

I still need to figure out how to properly track the desired calling convention (`UnmanagedFunctionPointer` is only for delegates right now, and we are technically emitting a struct) and how to differentiate an unmanaged function pointer from a managed function pointer.

But this is all just "proof of concept"/"prototype" right now anyways, so I'm not too worried.

**Comment by davidwrighton** [comment]
If a static delegate is wrapping a managed method, using the unmanaged calling convention will cause significant slowdowns. 

Also, thinking about memory management and lifetime, if it wraps a managed method, we really ought to have GC tracking such that the embedded function pointer has long-enough lifetime. This impacts things like AssemblyLoadContext and RunAndCollect reflection assemblies, as well as static delegates to LCG (dynamic) methods.

**Comment by davidwrighton** [comment]
The only way to hint to the runtime that something needs to be kept is to carry along some sort of object reference. Currently, an effective technique would be to use ldtoken to produce a RuntimeMethodHandle and carry that along (RuntimeMethodHandle is actually an object reference to an opaque object that will keep its associated method alive.) Unfortunately, I don't believe that will actually help any of these scenarios as simply needing to carry along a RuntimeMethodHandle is likely more expensive than using a normal delegate.

**Comment by jaredpar** [comment]
@davidwrighton can you elaborate on why this is more expensive? The allocation is amortized IIRC. Hence I'm confused where this would be more expensive.

**Comment by sunkin351** [comment]
So... How are we gonna tell if the function pointer in this static delegate is null? Can we get a method or property to determine this?

**Comment by GeirGrusom** [comment]
It's a value type so normal nullable value type semantics?

**Comment by Joe4evr** [comment]
I imagine it could be done via a computed property, similar to `ImmutableArray<T>`:
```cs
public bool IsDefault => _functionPtr == IntPtr.Zero;
```

**Comment by GeirGrusom** [comment]
Or just `staticDelegateField == default` though I think there is a semantic difference between "no value" and "pointer to zero".

**Comment by sunkin351** [comment]
I went ahead and implemented a placeholder for this idea, a struct wrapping a single `IntPtr` representing a function pointer, with runtime code generation backing it for it `Calli` opcode. (This creates a normal delegate, placed in a static field of the struct, so it will never be as fast as having the Calli inlined.) Works nicely. Even implemented a check in the IL for a null function pointer. I hope the .NET team moves forward with this Idea some time in the near future. I don't want to have to maintain this indefinitely.

**Comment by sunkin351** [comment]
Sooo... Anyone know if there's a chance this is gonna make it into C# 8.0?

**Comment by Joe4evr** [comment]
Since the Milestone is X.0, and the OP indicates there's been no decision on the proposal yet, I'm gonna go out on a limb and say "no".

**Comment by sunkin351** [comment]
According to this: https://github.com/dotnet/csharplang/tree/master/meetings/2019
It looks like they will be discussing "Function Pointers" on the 23rd of this month. To what end is the question.

**Comment by jaredpar** [comment]
@sunkin351 

> It looks like they will be discussing "Function Pointers" on the 23rd of this month. To what end is the question.

Mostly to finalize the proposal. The original proposal was well received a few months ago but there was some feedback on the syntax / naming rules. Those will be reviewed and implementation is likely to begin shortly there after, if not before. My hope is to have a PR out updating the proposal by EOW. 

Overall though my expectation is that if we move forward with function pointers that we will end up closing out this proposal. The main benefit of `static delegate` as proposed is that they're allocation free. The problem is that in an environment where assemblies can be unloaded there is really no way to have allocation free + safety. The `delegate` must have some type of strong handle back to the assembly which contains the method it refers to and that means it must have an allocation.

There were some designs where we felt like we could get the allocation down to one per target (create a `static delegate` for the same method 100 times and allocate only once). Those were a bit complex though and it didn't seem to justify the cost. 

In pretty much every case where we had a partner / customer who wanted `static delegate` the idea of a idea of using a function pointer + `struct` wrapper was an acceptable replacement. Hence it's likely the path forward for now.

**Comment by sunkin351** [comment]
I'll take function pointer's as long as I can make them point to managed methods without much trouble. All I really need them for is Native Interop. If assembly unloading is an issue, just make the developers using function pointers aware of it. They'll figure something out to work around the problem if it is one.

**Comment by NinoFloris** [comment]
> @davidwrighton can you elaborate on why this is more expensive? The allocation is amortized IIRC. Hence I'm confused where this would be more expensive.

Was any elaboration given for this?

What would prevent us from carrying around a struct with a pointer and a reference to the Type (or MethodHandle, or etc) containing the method we point at?
Shouldn't that be very fast, allocation free and prevent unloading?

**Comment by Thealexbarney** [comment]
@jaredpar From what I understand, you're saying that function pointers are more likely to happen, right? That improves allocations when using delegates, but it doesn't solve one of the other restrictions when using delegates in some cases. Are there any plans to address the following cases?

With the introduction of Span<T>, I'm seeing a lot of functions that used to use arrays move to using Spans instead. This can be great for performance, but it presents a problem when using closures. Currently closures over ref structs are not allowed. However with a "value delegate"-like feature it could be possible to use ref structs for closures.

The two obvious solutions I see are always passing the closure struct by reference, or storing references to the local variables in the closure. The second option would require support for ref fields in ref structs. The second option might also allow for capturing a function's ref parameters, something that's currently not allowed.

**Comment by jaredpar** [comment]
@Thealexbarney 

> From what I understand, you're saying that function pointers are more likely to happen, right?

Yes. At this point they're very likely going to be a part of C# 9.0

> Are there any plans to address the following cases?

Which cases are you speaking of?

> The second option would require support for ref fields in ref structs. The second option might also allow for capturing a function's ref parameters, something that's currently not allowed.

The `static delegate` feature though does not enable `ref struct` capture in lambdas. Nor does it make it easier to implement it via other means.  The difficulties around `ref struct` capture are still around how to thread the `ref struct` data through from the caller to the invoked method.

1. The heap can't be used because it's a `ref struct`
1. The closure can be passed as an explicit parameter to the delegate and threaded through the calling method (essentially think about `Enumerable.Where` taking the closure as a parameter and then passing it to the `predicate` argument). That is still very limiting because it must be typed to a single closure type as `ref struct` can't participate in generics

That takes most of the options off the table. 

The two language proposals that actually give a reasonable path for `ref struct` capture are:

1. `ref interface` #2975 
1. `where T : ref struct` #1148

Both of those at least allow parameterization around the type of the closure. It would still require non-trivial opt-in logic from the invoked function but it's doable at that point.

**Comment by Thealexbarney** [comment]
@jaredpar 
> Which cases are you speaking of?

Capturing byref-like types, and capturing `ref`s.

One example I've seen lately looks something like this. `Process` is a common function that will handle the delegate passed to it based on the additional arguments passed to `Process` and the program's internal state.

```cs
public static int Method1(Span<byte> data, param1, param2)
{
    return Process(() => Method1Internal(data, param1, param2), params_describing_how_function_should_be_handled);
}

private static int Method1Internal(Span<byte> data, param1, param2) { ... }
```

>That is still very limiting because it must be typed to a single closure type as ref struct can't participate in generics

I don't understand why usage of `ref struct`s as generic type parameters would affect `ref struct` capture. The `ref struct` would be a member of the closure struct, not a parameter type for a `Func<T>`.

Could a static delegate be implemented as a ref struct with a function pointer and an optional closure `ref struct` that would be used as the first parameter for that function?

**Comment by jaredpar** [comment]
@Thealexbarney 

> I don't understand why usage of ref structs as generic type parameters would affect ref struct capture. The ref struct would be a member of the closure struct, not a parameter type for a Func<T>.

I'm actually referring to the closure struct here, not the captured value. The captured value must be available to the method to which the `Func<>` refers. There are only limited ways to do this:

1. Attach the state to the `Target` member of the `System.Delegate` instance. This is how closure capture works today. That is simply not an option though once we get into the capture of `ref` or `ref struct` types because the `Target` parameter forces a heap allocation.
1. Pass the closure state as a argument to the `Func<>`. No heap capture is required here but it does mean that the `Func<>` in the signature needs to accept many different closure types. The only real options for doing for `ref` like capture is interfaces or generics

**Comment by Thealexbarney** [comment]
@jaredpar 
I thought a little about the details of how it might work by creating simplified examples.

This first one ended up not really working because any sort of struct inheritance I could think of wouldn't support both downcasting and overriding functions. A `ref interface` feature could work. From what I can tell that would require compiling a new version of `Foo` for every `ref interface` derived type it's called with, similar to generics.
```cs
// In source code
public static int LambdaUsingFunc(int a, Span<byte> span)
{
    int result = Foo(b =>
    {
        int value = b ? 4 : 8;
        a++;
        return value + a + span[0];
    });

    return result + a;
}

public static int Foo(ValueFunc<bool, int> func)
{
    return func.Invoke(true);
}

// As compiled by Roslyn
public static int LambdaUsingFunc(int a, Span<byte> span)
{
    var closure = new Closure();
    closure._a = a;
    closure._span = span;

    var funcStruct = new FuncStruct();
    funcStruct.FuncPtr = &Closure.Lambda_1; // Or however you create a function pointer.
    funcStruct.Closure = ref closure;

    int result = Foo(funcStruct);
    
    return result + closure._a;
}

[CompilerGenerated]
internal ref struct Closure
{
    public int _a;
    public Span<byte> _span;

    public static int Lambda_1(Closure closure, bool b)
    {
        int value = b ? 4 : 8;
        closure._a++;
        return value + closure._a + closure._span[0];
    }
}

// I'm not sure how this could work since structs can't inherit structs and don't have vtables
[CompilerGenerated]
internal ref struct FuncStruct : ValueFunc<bool, int>
{
    public delegate*<Closure, bool, int> FuncPtr;
    public ref Closure Closure;       // Requires ref fields

    public int Invoke(bool a)
    {
        return FuncPtr(Closure, a);
    }
}
```

> Pass the closure state as a argument to the Func<>. No heap capture is required here but it does mean that the Func<> in the signature needs to accept many different closure types. The only real options for doing for ref like capture is interfaces or generics

Do you mean something like this?
```cs
// This could work with ref fields and ref struct constraints, but it makes the signatures for
// functions using a ValueFunc more complicated and could cause generic bloat
internal ref struct ValueFunc<TClosure, TIn, TOut> where TClosure : ref struct
{
    public delegate*<TClosure, TIn, TOut> FuncPtr;
    public ref TClosure Closure;       // Requires ref fields

    public TOut Invoke(TIn a)
    {
        return FuncPtr(Closure, a);
    }
}

public static int Foo<TClosure>(ValueFunc<TClosure, bool, int> func)
{
    return func.Invoke(true);
}
```
Using `IntPtr` might work like in the original `static delegate` proposal. The struct would be created by the compiler, so it could verify the ValueFunc type parameters, FuncPtr and Closure types all agree with each other.
```cs
// IntPtrs could be used. The compiler would make sure the types they point to are valid
public ref struct ValueFunc<bool, int>
{
    private IntPtr FuncPtr;
    private IntPtr Closure; // Points to a ref struct. GC wouldn't be an issue because 
                            // the struct must live somewhere else on the stack.

    public int Invoke(bool a)
    {
        // Make sure FuncPtr isn't null
        
        // If Closure isn't null, call FuncPtr with Closure as the first argument,
        // otherwise call FuncPtr without closure
    }
}
```

**Comment by weltkante** [comment]
But what would you do with such a delegate containing a ref-struct closure? You couldn't store it anywhere so its basically just a glorified inversion-of-control scheme to write your algorithms differently.

Not saying thats a useless thing but wanting to point at the fact that this is an entirely different kind of delegate incompatible to anything currently operating on `Action` or `Func` delegates, you'd have to write entirely new infrastructure to operate on those delegates.

**Comment by Thealexbarney** [comment]
@weltkante You'd use it in places where the delegate is only used during execution of the function it's passed to. This is fairly common in my experience. I gave an example a couple posts back, and I've seen people discussing its use for making more optimized LINQ-like functions.

**Comment by jaredpar** [comment]
@Thealexbarney 

> // I'm not sure how this could work since structs can't inherit structs and don't have vtables

This is the crux of the problem with this approach. It's not workable because a `struct` can't have a vtable. In order to give a `struct` implementation defined behavior it essentially needs to either use an interface (explicit way to provide a v-table) or generics. That's why I listed those two particular issues as the real blockers for `ref` based capture.

**Comment by Sergio0694** [comment]
> I gave an example a couple posts back, and I've seen people discussing its use for making more optimized LINQ-like functions.

@Thealexbarney Can't you already have LINQ-like methods acting on `Span<T>`-like types, as long as you define your own `ref struct` enumerators that go along with them? [`Hyperlinq`](https://github.com/NetFabric/NetFabric.Hyperlinq) is doing exactly this: it exposes extensions for `Span<T>` and `ReadOnlySpan<T>` similar to LINQ APIs, taking predicates and whatnot, then captures them in a `ref struct` enumerator and uses that to produce lazy results over the initial sequence. You're basically suggesting a way to achieve this without having to manually define your own `ref struct` enumerators every time, correct?

**Comment by tannergooding** [comment]
> It's not workable because a struct can't have a vtable

Mostly joking (this would be incredibly unsafe/complex, etc) but you could always emulate a vtbl. A vtbl in C/C++ is generally just (yes this is a COM like example, but it applies to both the MSVC++ ABI used by Windows and the Itanium C++ ABI used by Linux/macOS as well):
```csharp
struct IUnknown {
    IUnknownVtbl* lpVtbl;
};

struct IUnknownVtbl {
    public delegate*<IUnknown*, Guid*, void**, int> QueryInterface;
    public delegate*<IUnknown*, uint> AddRef;
    public delegate*<IUnknown*, uint> Release;
};
```

where the constructor for the "type" initializes the function pointers to the right addresses. The same principal could be applied to other cases where you want to have pseudo inheritance with structs.

**Comment by jaredpar** [comment]
@tannergooding 

Yes but once again: how do you pass the closure around? Your example is a valid implementation of a COM v-table but it lacks any inherently associated data.

The crux of this problem remains the same: how to get the closure from the frame it's created to the frame it's used. I've laid out the available options I'm aware of.

**Comment by tannergooding** [comment]
Assuming this is a ref struct and the closure is only be passed down (and never stored elsewhere). It should be fine to just create a `ref struct Closure` and then store it alongside the function pointer as a `T*` (stored in a `ref struct ValueDelegate`) to the location on the stack.
It isn't "safe", but provided you follow the storage rules for a `ref struct`, there shouldn't be an issue functionally speaking as the `Closure` will live the entire time that the `ValueDelegate` is to live and not longer (and this will then just be working around https://github.com/dotnet/runtime/issues/32060 not existing yet).

**Comment by jaredpar** [comment]
> It should be fine to just create a ref struct Closure and then store it alongside the function pointer as a T* (stored in a ref struct ValueDelegate) to the location on the stack.

That's not fine though because it leaves a GC gap. Assuming the closure isn't used after the method invocation, which is a decent bet, then the GC will consider all of the data within the closure to be collectable. The `T*` isn't a strong reference hence it won't help here.

**Comment by tannergooding** [comment]
Ah right. I was assuming the Closure was `unmanaged` which may not be the case.

**Comment by Thealexbarney** [comment]
@Sergio0694 Struct enumerators would be a different thing. I recently saw some people talking about how a `static delegate`-like concept would help with performance for operations that have a function as a parameter, but using `yield return` would still create a class.

@jaredpar That wouldn't be an issue as long as the lifetime of the closure is at least as long as the `ValueDelegate`. Something like `KeepAlive` could be used if needed.

**Comment by jaredpar** [comment]
@Thealexbarney 

> that wouldn't be an issue as long as the lifetime of the closure is at least as long as the ValueDelegate. Something like KeepAlive could be used if needed.

That introduces two new technologies:

1. A `KeepAlive` that is usable for any arbitrary `ref struct`. Such a function does not exist and is not really authorable unless we implement the `ref struct` support for generics or interfaces (again the two issues I listed above). 
2. In order to get the lifetime guarantee that you specified the `delegate` type we pass down must be a `ref struct` as well. That is also not available today. 

Even assuming those two existed and those problems were solved @tannergooding sample is still lacking. How do the dots get connected from the point the closure is created to the point that the closure is used? His comments essentially suggest strong typing in the form of `Closure*`. Again though that doesn't work because every closure has a different type meaning the containing "value delegate" would also need a different type. To really make this work you essentially need to have the following setup:

```c#
unsafe ref struct RefStructDelegate {
  void* Closure;
  int Func<void*, int> Target;
```

Then you can essentially have the following pattern:

```c#
void Producer() {
  Closure$ c = default;
  c.span = ...;
  RefStructDelegate d = default;
  d.Closure = &c;
  d.Target = Lambda$;
  Consumer(d);
  KeepAlive(c); // Somehow
}

unsafe int Consumer(RefStructDelegate d) {
   return d.Target(d.Closure, 42);
}

unsafe int Lambda$(void* pClosure, int data) { 
  ref Closure c = Unsafe.ThatMethodWhichMakesPointersRefs((Closure*)pClosure));
  return ... 
```

This has the further downside that the consumer must use `unsafe` code because the `RefStructDelegate` is itself `unsafe`. That really defeats the purpose of using `Span<T>` in the firs place. The language is much more likely going to prefer finding safe solutions to these problems vs. unsafe ones.

**Comment by jvbsl** [comment]
Is there somewhere one can read up on the reason why this was rejected, and which alternatives are in discussion(if any)?
I would like to know why something so important to platform independency for bigger things than "Hello World" native interop was rejected.
I was pretty happy when this was discussed for 9.0 and hoped to have it with .Net 5 for implementation of OpenTK. As it is currently impossible to have a library be able to do native symbol resolving if a custom symbol loader is provided by the C interface. Which is impossible to avoid with OpenGL/OpenAL and Vulkan. It would be really nice to have more control over low level stuff like that, or possibility to use DllImport with custom symbol resolving. Now the only possible ways are writing code in IL or using Marshal.GetDelegateForFunctionPointer which is as far as I'm aware still a good bit worse on performance than calli, or DllImport.

**Comment by weltkante** [comment]
I think the [function pointers proposal](https://github.com/dotnet/csharplang/blob/master/proposals/function-pointers.md) overlaps with this so it makes no sense to implement this one as well? Work seems to [have started](https://github.com/dotnet/roslyn/issues/39865) on function pointers already.

**Comment by 333fred** [comment]
That is correct, this proposal was superseded by function pointers.

**Comment by nxrighthere** [comment]
Uh, it would be great if we do some cleanup/closing/tagging with such proposals to track the appropriate ones instead.

**Comment by jaredpar** [comment]
@nxrighthere once function pointers has actually shipped this proposal will get closed out. Generally we don't close the main proposal or competing ones until the features have shipped.

**Comment by nxrighthere** [comment]
Got it, thanks. I think GitHub is really missing a sort of "issues chains/dependencies" or a feature like that to explicitly notice such cases.

**Comment by jvbsl** [comment]
Then I was upset for nothing :D Thanks for the clarification

**Comment by NetMage** [comment]
Shouldn't the initial post be updated as Finalized and perhaps a line added pointing to the superceding feature as the reason?

**Comment by jaredpar** [comment]
Once we ship yes.

**Comment by 333fred** [comment]
Superseded by https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md. Closing out.
